src/v/storage/disk_log_impl.cc

```diff
@@ -2313,4 +2314,201 @@ void disk_log_impl::set_cloud_gc_offset(model::offset offset) {
+          config().ntp());
+        co_return res;
+    }
+
+    /*
+     * calculate the effective local retention. this forces the local retention
+     * override in contrast to housekeeping GC where the overrides are applied
+     * only when local retention is non-advisory.
+     */
+    cfg = apply_base_overrides(cfg);
+    cfg = override_retention_config(cfg);
+    const auto local_retention_offset
+      = co_await maybe_adjusted_retention_offset(cfg);
+
+    /*
+     * when local retention is based off an explicit override, then we treat it
+     * as the partition having a retention hint and use it to deprioritize
+     * selection of data to reclaim over data without any hints.
+     */
+    const auto hinted = has_local_retention_override();
+
+    /*
+     * for a cloud-backed topic the max collecible offset is the threshold below
+     * which data has been uploaded and can safely be removed from local disk.
+     */
+    const auto max_collectible = stm_manager()->max_collectible_offset();
+
+    /*
+     * lightweight segment set copy for safe iteration
+     */
+    fragmented_vector<segment_set::type> segments;
+    for (const auto& seg : _segs) {
+        segments.push_back(seg);
+    }
+
+    /*
+     * currently we use two segments as the low watermark size
+     */
+    std::optional<model::offset> low_watermark_offset;
+    if (segments.size() >= 2) {
+        low_watermark_offset
+          = segments[segments.size() - 2]->offsets().base_offset;
+    }
+
+    vlog(
+      stlog.debug,
+      "Categorizing {} {} segments for {} with local retention {} low "
+      "watermark {} max collectible {}",
+      segments.size(),
+      (hinted ? "hinted" : "non-hinted"),
+      config().ntp(),
+      local_retention_offset,
+      low_watermark_offset,
+      max_collectible);
+
+    /*
+     * categorize each segment.
+     */
+    for (const auto& seg : segments) {
+        const auto usage = co_await seg->persistent_size();
+        const auto seg_size = usage.total();
+
+        /*
+         * the active segment designation takes precedence because it requires
+         * special consideration related to the implications of force rolling.
+         */
+        if (seg == segments.back()) {
+            /*
+             * since the active segment receives all new data at any given time
+             * it may not be fully uploaded to cloud storage so it is hard to
+             * say anything definitive about it. instead, we report its size as
+             * a potential:
+             *
+             *   1. rolling the active segment will bound progress towards
+             *   making its current full size reclaimable as max collectible
+             *   increases to cover the entire segment.
+             *
+             *   2. finally, we don't report it if max collectible hasn't even
+             *   made it to the active segment yet.
+             */
+            if (seg->offsets().base_offset <= max_collectible) {
```

Comment:
thought: I think now it is guaranteed that max_collectible lies at segment boundaries but if in the future for some reason it didn't this logic would have to change (i think). Maybe it would be good enough to also check that the segment's latest offset is also less then max collectible

Reply:
Yeah that makes a lot of sense +1